### PR TITLE
One common Tree object per application

### LINKED ISSUE #1242: TreeView should update on "create" "delete" children etc.

### LINKED ISSUE #1243: ONE tree instance on the front-end

### PR DESCRIPTION
**Pull request by soofstad**
## What does this pull request change?
- Setup a "FSTreeContext" with shared Tree state for each application
- Tree is passed "setIndex" so it can trigger rerenders based on any updates to the trees state
- Add/Delete nodes are now correctly updated in the view without expanding manually
- Each instance of TreeView keeps its own state of expanded and minimized nodes.

## Issues related to this change
closes #1242 closes #1243

### CODE REVIEW

File: web/packages/plugins/apps/dmt-app/src/pages/editor/Editor.tsx [hunk 1]
```diff
@@ -73,6 +79,7 @@ export default () => {
       <div style={{ display: 'flex' }}>
         <TreeWrapper>
           <TreeView
+            index={index}
```
Comment (KristianKjerstad) [inline web/packages/plugins/apps/dmt-app/src/pages/editor/Editor.tsx:82]: same as previous comment, I was confused about the name "index"

File: web/packages/plugins/shared/common/src/components/TreeView.tsx [hunk 2]
```diff
@@ -117,72 +118,56 @@ const TreeNodeComponent = (props: {
...
-      .init()
-      // @ts-ignore
-      .then(() => setIndex([...tree]))
-      .finally(() => setLoading(false))
+    let expandedNodes: { [k: string]: boolean } = {}
+    index.forEach((node: TreeNode) => {
+      // Initialize expanded state where only top level DataSources are expanded
+      expandedNodes[node.nodeId] = node.isDataSource
+    })
+    setExpandedNodes(expandedNodes)
```
Comment (KristianKjerstad) [inline web/packages/plugins/shared/common/src/components/TreeView.tsx:139]: It looks like this part should expand the datasources, but this is what I get when I refresh the page (nothing is expanded)

image

however, the tree in the EntityPicker for example seems to be expanded correctly

**Review by KristianKjerstad — CHANGES_REQUESTED** [review]
nice!

just some small comments

File: web/packages/plugins/shared/common/src/context/FileSystemTreeContext.tsx [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+}>
+// @ts-ignore
+FSTreeContext = createContext({ tree: null, index: [], loading: false })
+
+export const FSTreeProvider = (props: { children: ReactNode }) => {
+  const { children } = props
+  const [loading, setLoading] = useState<boolean>(true)
+  const { token } = useContext(AuthContext)
+  const appConfig = useContext(ApplicationContext)
+  const [index, setIndex] = useState<TreeNode[]>([])
```
[the remark this answers is not in the document]
Comment (soofstad) [reply web/packages/plugins/shared/common/src/context/FileSystemTreeContext.tsx:23]: Agreed. Renamed

File: web/packages/plugins/shared/common/src/components/TreeView.tsx (hunk 2 above)
Comment (soofstad) [reply web/packages/plugins/shared/common/src/components/TreeView.tsx:139]: Good catch. Fixed by waiting for the loading to be complete before rendering the tree

**Pull request merged**